build-logic-settings/build-environment/src/main/kotlin/gradlebuild.build-environment.settings.gradle.kts

```diff
@@ -20,11 +20,31 @@ import gradlebuild.basics.BuildEnvironmentService
 with(layout.rootDirectory) {
     gradle.lifecycle.beforeProject {
         val service = gradle.sharedServices.registerIfAbsent("buildEnvironmentService", BuildEnvironmentService::class) {
+            assert(project.path == ":") {
+                // We rely on the fact that root is configured first
+                "BuildEnvironmentService should be registered by the root"
+            }
             parameters.rootProjectDir = this@with
+            parameters.rootProjectBuildDir = project.layout.buildDirectory
+            parameters.artifactoryUserName = providers.gradleProperty("artifactoryUserName")
+            parameters.artifactoryPassword = providers.gradleProperty("artifactoryPassword")
+            parameters.testVersions = providers.gradleProperty("testVersions")
+            parameters.integtestAgentAllowed = providers.gradleProperty("org.gradle.integtest.agent.allowed")
+            parameters.integtestDebug = providers.gradleProperty("org.gradle.integtest.debug")
+            parameters.integtestLauncherDebug = providers.gradleProperty("org.gradle.integtest.launcher.debug")
+            parameters.integtestVerbose = providers.gradleProperty("org.gradle.integtest.verbose")
```

Comment:
🤔  Using `providers.gradleProperty` should not be an IP violation.

It seems we went straight from `project.hasProperty("testVersions")` to using the extension. Can we try replacing them with `providers.gradleProperty("testVersions")` at the use-sites first?

Reply:
Was this improved in latest IP detection? Because I believe `providers.gradleProperty` was detected as a read of a property on `:` project.

Reply:
I am not aware if it was ever something that resolved against the root project.

It seems like from the very beginning it was a `ValueSource`, all of which are IP-safe:
https://github.com/gradle/gradle/blob/0718ca327fe076ae4ec56e2076999bf3cc93be24/platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/DefaultProviderFactory.java#L135-L146

Reply:
Fixed, now I use `providers.gradleProperty`